Fail the status check when the coverage decreased

lextiz opened the pull request:
This is feature is essential for enforcing the code coverage, using GH status and branch protection.

dmotpan commented:
@lextiz I think that this feature should be customizable as there are cases when the total coverage decreases because of some removed classes and their tests and the build should not fail. 
Also, this logic should be covered with a test.

jaredj commented:
FWIW github allows you to configure whether a failed status check prevents merge or not.  If you don't want coverage reductions to interrupt anything you can just turn that off and let the coverage status "failure" just make it more obvious when coverage was reduced

lextiz commented:
@dmotpan Agreed. This will take a while.

KenneyE commented:
Sorry, will this PR be moving forward? It would be helpful. I agree with @jaredj that someone could always mark the check as not required, and as the point of this plugin is to compare coverage across branches and report the result on GH, it seems like clearly indicating the result would be helpful to most users.

dmotpan commented:
@KenneyE anyway, a unit test is required for the logic introduced in this PR

lextiz commented:
I have invested some time in writing the unit tests for this PR and would like to ask for some advice to proceed:
1. The unit tests require to mock a static method (namely ```get``` method of ```ServiceRegistry``` ) to emulate different code coverage percentage for master and the PR branches. Mockito [does not support](https://github.com/mockito/mockito/issues/1013) mocking of such methods. Is that OK to add [PowerMock](https://github.com/powermock/powermock) as an additional test scoped dependency?
1. As @jaredj mentioned GH allows to define a check as not required (this is actually a default). Would this PR be accepted without adding the "do not fail the GH check" flag?

dmotpan commented:
@lextiz 
1. Can't we make use of `set` methods from `ServiceRegistry` and set mocks to it as it is done in `CompareCoverageActionTest`? So that the get method always return a mock. It's done this way in `CompareCoverageActionTest`
2. Yes, it would be accepted

lextiz commented:
Ready for CR.

dmotpan commented on src/test/java/com/github/terma/jenkins/githubprcoveragestatus/CompareCoverageActionTest.java:
```diff
@@ -82,20 +88,36 @@ public void postCoverageStatusToPullRequestAsComment() throws IOException, Inter
+        prepareEnvVars();
+        prepareCommit();
+        prepareCoverageData(0.95f, 0.9f);
+        coverageAction.setPublishResultAs("statusCheck");
```
let's separate each test step (given/when/then approach) with a new line

dmotpan commented on src/test/java/com/github/terma/jenkins/githubprcoveragestatus/CompareCoverageActionTest.java:
```diff
@@ -82,20 +88,36 @@ public void postCoverageStatusToPullRequestAsComment() throws IOException, Inter
         prepareCommit();
+        prepareCoverageData(0.88f, 0.95f);
         coverageAction.setPublishResultAs("statusCheck");
-
```
please bring back the new lines in order to keep given/when/then test style

lextiz commented:
Thanks for quick review, I have corrected the remarks.

jaredj commented:
just to play devil's advocate with myself... one use case for wanting to make this a "required" check but still be able to merge in spite of coverage reduction, would be if a team wants to *prevent merging when the check has not yet been completed*.   If this is an important use case, then an option to never fail the check might in fact be necessary.

If someone did add an option, it might be useful to make that option a float e.g. representing how much coverage can be reduced before a failure is triggered.  e.g. `0.0` would be "always fail if coverage is reduced", `1.0` would be "fail if coverage is reduced by >= 1%", and perhaps `-1.0` would be "fail if coverage is not increased by >= 1%"

lextiz commented:
@dmotpan Is there something else that needs to be done on my side to finalize the PR?

dmotpan commented on src/test/java/com/github/terma/jenkins/githubprcoveragestatus/CompareCoverageActionTest.java:
```diff
@@ -82,20 +88,40 @@ public void postCoverageStatusToPullRequestAsComment() throws IOException, Inter
         prepareCommit();
+        prepareCoverageData(0.88f, 0.95f);
         coverageAction.setPublishResultAs("statusCheck");
-
```
@lextiz just remove the unnecessary spaces in the empty line here and below. I am going to test the PR and merge if all good

dmotpan commented:
@lextiz have you tested it? I am seeing a success status check even if the coverage decreased
![image](https://user-images.githubusercontent.com/17102131/54704965-738d2500-4b44-11e9-98cf-a0adaaa447c9.png)

lextiz commented:
@dmotpan 
Yes, I have added 2 tests: one for successful PR status and one for failed and removed one that did not change the coverage percentage. Now I have brought back the original test, lets see if it will fix the coverage. Is there a detailed (per line) coverage report, so I could fix the coverage decrease? I cannot see the statuses presented at the screenshot, but only the "continuous-integration/jenkins/pr-merge" which is green.

lextiz commented:
I did not do end-to-end test, assuming existing status assertions in the unit tests are working. How did you test it? Could we maybe have a short call to finalize the test?

dmotpan commented:
@lextiz I found the issue - the implementation of `createCommitStatus` [always uses](https://github.com/jenkinsci/github-pr-coverage-status-plugin/blob/master/src/main/java/com/github/terma/jenkins/githubprcoveragestatus/GitHubPullRequestRepository.java#L93) `SUCCESS` instead of the `state` passed to the method. We should use `state` instead of success and the PR should be good to merge

dmotpan commented on src/test/java/com/github/terma/jenkins/githubprcoveragestatus/CompareCoverageActionTest.java:
```diff
@@ -80,7 +86,7 @@ public void postCoverageStatusToPullRequestAsComment() throws IOException, Inter
         verify(pullRequestRepository).comment(ghRepository, 12, "[![0% (0.0%) vs master 0%](aaa/coverage-status-icon/?coverage=0.0&masterCoverage=0.0)](aaa/job/a)");
     }
-
+    
```
please remove the spaces

lextiz commented:
@dmotpan Thank you for help! Done.

dmotpan commented:
@lextiz thanks for contribution

KenneyE commented:
This is great and will make this plugin even more helpful. Thank you! 
I'm not an expert at Jenkins. Is there any way I can update my plugin to get this change, or do I need to wait for a version bump?

Thanks again!

chris-cayaba commented:
> @lextiz have you tested it? I am seeing a success status check even if the coverage decreased

Hi I am still getting this. help please?

chris-cayaba commented:
Hello,
![image](https://user-images.githubusercontent.com/32443852/64139457-903bc180-ce33-11e9-8a9b-a24d9a1257d9.png)

I'm still getting a success although the coverage is negative/reduced.

dmotpan commented:
@chris-cayaba the version is not yet released and this is why you still seet

chris-cayaba commented:
> @chris-cayaba the version is not yet released and this is why you still see the success mark

Hi @dmotpan
Ok, thank you.

lexia-wouter-vanrij commented:
I see this pull request is already merged, but not yet released. For when is the next release scheduled?

jordanjennings commented:
Is there any way to disable or tune this feature? We have a variety of cases where a very slight decrease in coverage is acceptable, for example a PR that deleted some code and went down 0.2% in coverage now fails the check.
